invenio_github/views/github.py

```diff
@@ -24,225 +24,221 @@
+    """
     user_id = current_user.id
     github = GitHubAPI(user_id=user_id)
-    token = github.session_token
-
-    if token:
-        repos = github.account.extra_data.get('repos', [])
-        repo = next((repo for repo_id, repo in repos.items()
-                     if repo.get('full_name') == name), {})
-        if not repo:
-            abort(403)
-
-        try:
-            # NOTE: Here we do not check for repository ownership, since it
-            # might have changed even though the user might have made releases
-            # in the past.
-            repo_instance = Repository.get(user_id=user_id,
-                                           github_id=repo['id'],
-                                           check_owner=False)
-        except RepositoryAccessError:
-            abort(403)
-        except NoResultFound:
-            repo_instance = Repository(name=repo['full_name'],
-                                       github_id=repo['id'])
-
-        releases = [
-            current_github.release_api_class(r) for r in (
-                repo_instance.releases.order_by(db.desc(Release.created)).all()
-                if repo_instance.id else []
-            )
-        ]
+
+    try:
+        # NOTE: Here we do not check for repository ownership, since it
+        # might have changed even though the user might have made releases
+        # in the past.
+        repo = github.get_repository(repo_name)
+        releases = github.get_repository_releases(repo=repo)
         return render_template(
-            current_app.config['GITHUB_TEMPLATE_VIEW'],
-            repo=repo_instance,
-            repo_info=repo,
+            current_app.config["GITHUB_TEMPLATE_VIEW"],
+            repo=repo,
             releases=releases,
             serializer=current_github.record_serializer,
         )
+    except RepositoryAccessError as e:
+        abort(403)
+    except NoResultFound as e:
+        abort(404)
 
-    abort(403)
 
+###
+# TODO to be moved to its own folder (e.g. separate ui / api)
+# /api routes
+###
 
-@blueprint.route('/rejected')
-@login_required
-def rejected():
-    """View for when user rejects request to connect to github."""
-    return render_template('invenio_github/settings/rejected.html')
+blueprint_api = Blueprint("invenio_github_api", __name__)
 
 
-@blueprint.route('/hook', methods=['POST', 'DELETE'])
 @login_required
-def hook():
-    """Install or remove GitHub webhook."""
-    repo_id = request.json['id']
+@blueprint_api.route("/user/github/repositories/sync", methods=["POST"])
+def sync_user_repositories():
+    """Synchronizes user repos.
+
+    Currently:
+        POST /api/user/github/repositories/sync
+    Previously:
+        POST /account/settings/github/hook
+    """
+    try:
+        github = GitHubAPI(user_id=current_user.id)
+        github.sync(async_hooks=False)
+        db.session.commit()
+    except Exception:
+        db.session.rollback()
```

Comment:
I moved away from nested sessions, as they were giving me some trouble. Namely, when we have to commit a session before spawning a celery task it would make the nested session error on exit.